crates/dwallet-mpc-types/src/dwallet_mpc.rs

```diff
@@ -31,6 +32,101 @@ pub const START_NETWORK_DKG_EVENT_STRUCT_NAME: &IdentStr =
+    pub number_of_chunks: Option<usize>,
+}
+
+/// Represents the state of the message building process.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum MessageState {
+    /// All chunks received; full message reconstructed.
+    Complete(MPCMessage),
+    /// Still waiting on some chunks; maps sequence numbers to slices.
+    Incomplete(HashMap<u64, MPCMessageSlice>),
+}
+
+/// Builds and reconstructs messages from incoming slices.
+/// Some MPC messages might be greater than the maximum size of a consensus message limit.
+/// The `MPCMessageBuilder` is used to split the message into smaller chunks,
+/// to avoid exceeding the maximum size limit.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct MPCMessageBuilder {
+    /// Current state of the message.
+    pub messages: MessageState,
+}
+
+impl MPCMessageBuilder {
+    pub fn empty() -> Self {
+        Self {
+            messages: MessageState::Incomplete(HashMap::new()),
+        }
+    }
+
+    /// Splits a message into smaller chunks and returns a new builder with those chunks.
+    /// Ensures at least one chunk is created, even if the message is empty.
+    pub fn split(message: Vec<u8>, chunk_size: usize) -> Self {
+        let chunks: Vec<Vec<u8>> = if message.is_empty() {
+            vec![vec![]] // Ensure at least one slice for empty messages
+        } else {
+            message
+                .chunks(chunk_size)
+                .map(|chunk| chunk.to_vec())
+                .collect()
+        };
+
+        let number_of_chunks = chunks.len();
+
+        let messages = chunks
+            .into_iter()
+            .enumerate()
+            .map(|(i, message)| {
+                (
+                    i as u64,
+                    MPCMessageSlice {
+                        message,
+                        sequence_number: i as u64,
+                        number_of_chunks: if i == 0 { Some(number_of_chunks) } else { None },
+                    },
+                )
+            })
+            .collect();
+
+        Self {
+            messages: MessageState::Incomplete(messages),
+        }
+    }
+
+    /// Adds a message slice to the builder and attempts to complete the full message.
+    /// If all slices are present, the message state is updated to `Complete`.
+    pub fn add_and_maybe_complete(&mut self, message: MPCMessageSlice) {
+        if let MessageState::Incomplete(messages) = &mut self.messages {
+            messages.insert(message.sequence_number, message);
+
+            if let Some(slice) = messages.get(&0) {
+                let expected_chunks = slice.number_of_chunks.unwrap_or(0);
+
+                if messages.len() == expected_chunks {
+                    let complete_message = (0..expected_chunks as u64)
+                        .map(|i| messages.get(&i).map(|s| s.message.as_slice()))
+                        .collect::<Option<Vec<_>>>()
+                        .map(|slices| slices.concat());
+
```

Comment:
`concat` is `[nighlty](https://doc.rust-lang.org/std/slice/trait.Concat.html)`, we can't use it. 
```suggestion
                    let complete_message = (0..expected_chunks as u64)
                        .map(|i| messages.get(&i))
                        .collect::<Option<Vec<_>>>()
                        .map(|slices| slices.into_iter().flatten().collect::<Vec<_>>());

```